src/box/relay.cc

```diff
@@ -564,13 +564,16 @@ static void
 tx_status_update(struct cmsg *msg)
 {
 	struct relay_status_msg *status = (struct relay_status_msg *)msg;
-	vclock_copy(&status->relay->tx.vclock, &status->vclock);
-	status->relay->tx.txn_lag = status->txn_lag;
-
+	struct relay *relay = status->relay;
+	if (vclock_get(&relay->tx.vclock, instance_id) !=
+	    vclock_get(&status->vclock, instance_id)) {
```

Comment:
1. Why not `vclock_compare`? Shouldn't we update the lag whenever any vclock component changes?
2. Can we skip update of `relay->txn_lag` in this case? I don't quite like that now `relay->txn_lag` and `relay->tx.txn_lag` can be different. 
3. It'd be nice to have a brief comment here explaining why you skip the lag update.

Reply:
1. No. The downstream lag is only about changes authored by this master. And comparing by vclock will lead to one of the cases shown in my test: when another instance writes something, downstream lag gets increased.
2. I guess we can. It was skipped always when replica vclock is unchanged before commit https://github.com/tarantool/tarantool/commit/56571d83172f991d2e8c09b85c27f079c067fe43.
You see, replica keeps sending last applied row timestamp in its acks even when there are no updates. And each time the ack arrives, relay->txn_lag is updated with `now() - applied row timestamp`. So it grows indefinitely in relay when there are no new rows.
3. Sure.

Thanks for asking! I came up with a better fix while trying to formulate the answers to your questions.
Here it is (together with an explanation comment). I haven't updated the branch yet, will do once I fix Borya's comments.
<details><summary>diff</summary>

```diff
diff --git a/src/box/relay.cc b/src/box/relay.cc
index f2f4aa4d4..388ca6229 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -564,16 +564,13 @@ static void
 tx_status_update(struct cmsg *msg)
 {
        struct relay_status_msg *status = (struct relay_status_msg *)msg;
-       struct relay *relay = status->relay;
-       if (vclock_get(&relay->tx.vclock, instance_id) !=
-           vclock_get(&status->vclock, instance_id)) {
-               relay->tx.txn_lag = status->txn_lag;
-       }
-       vclock_copy(&relay->tx.vclock, &status->vclock);
+       vclock_copy(&status->relay->tx.vclock, &status->vclock);
+       status->relay->tx.txn_lag = status->txn_lag;
+
        struct replication_ack ack;
-       ack.source = relay->replica->id;
+       ack.source = status->relay->replica->id;
        ack.vclock = &status->vclock;
-       bool anon = relay->replica->anon;
+       bool anon = status->relay->replica->anon;
        /*
         * Let pending synchronous transactions know, which of
         * them were successfully sent to the replica. Acks are
@@ -591,7 +588,7 @@ tx_status_update(struct cmsg *msg)
                {relay_status_update, NULL}
        };
        cmsg_init(msg, route);
-       cpipe_push(&relay->relay_pipe, msg);
+       cpipe_push(&status->relay->relay_pipe, msg);
 }
 
 /**
@@ -720,9 +717,14 @@ relay_reader_f(va_list ap)
                         * been written to WAL with our current realtime
                         * clock value, thus when it get reported back we
                         * can compute time spent regardless of the clock
-                        * value on remote replica.
+                        * value on remote replica. Update the lag only when the
+                        * timestamp corresponds to some transaction the replica
+                        * has just applied, i.e. received vclock is bigger than
+                        * the previous one.
                         */
-                       if (xrow.tm != 0)
+                       if (xrow.tm != 0 &&
+                           vclock_get(&relay->status_msg.vclock, instance_id) <
+                           vclock_get(&relay->recv_vclock, instance_id))
                                relay->txn_lag = ev_now(loop()) - xrow.tm;
                        fiber_cond_signal(&relay->reader_cond);
                }

```

</details>